### DISCUSSION TITLE
C# monkey patching / concrete members overriding

### DISCUSSION THREAD

**Discussion by przemo098** [dotnet/csharplang#8017] [Language Ideas]
Hello

I want to propose well known monkey patching 
I think it wont be used often in app code, however it will simplify writing tests / overriding sealed classes in libs
What is really important in development lifecycle.
I know there are libraries like Harmony / Microsoft Fake / MockType, however all of them are commercial and cumbersome.
It's also possible with Castle Windsor / Mock libs however in case of overriding exisitng library / nuget it's impossible with concrete not virtual class members

Usage:

```
public class Calculator{
    public int Add(int a, int b){
        return a + b;
    }
}
```

Test.cs
```
var calc = new Calculator();
calc.Add = (int a, int b) => a * b;
calc.Add(4,4);  // return 16 instead 8
```

This effect can be achieved in different manors, maybe even introduction of new keyword like **force** would be enough

```
public class CalculatorProxy : Calculator{
    public **force** int Add(int a, int b){
        return a * b;
    }
}
```

Program.cs
```
Calculator calc = new CalculatorProxy();
calc.Add(4,4);   // expected 16 instead of 8
```

**Comment by HaloFour** [comment]
I don't even think that this would be possible from within the language itself.  You'd effectively have to rewrite the target assembly, which is well outside of what the compiler can or should do.

**Reply by przemo098** [reply]
I think compiler can check this code and create syntactic sugar where it will create additional classes for this use case,
I know it's not trivial :), but I believe there are more than 1 way to solve it

**Reply by CyrusNajmabadi** [reply]
> and create syntactic sugar where it will create additional classes for this use case,

As part of this idea, can you flesh this out so we can get an idea of what sort of solution would work here?  Thanks!

**Reply by przemo098** [reply]
I think compiler should traverse (inheritance) classes which have **force** keyword used and at this point mark parent members as virtual ?

**Reply by CyrusNajmabadi** [reply]
@przemo098 The compiler can't do that.  The types are baked into dlls.  If they're types from your own source, then you could trivially make the methods virtual yourself.

**Reply by przemo098** [reply]
If it's in external dll than compiler can wrap them and create new dll based on existing
Also it can handle differently **force** keyword in case of casting
Also monkey patching was implemented in python / javascript

**Reply by CyrusNajmabadi** [reply]
> Also monkey patching was implemented in python / javascript

Python and JS are fundamentally different (and dynamic-forward) runtimes.  They are designed around the idea that these mutations are supported at a runtime level.  That's not the case for the .Net runtime.

**Reply by CyrusNajmabadi** [reply]
> If it's in external dll than compiler can wrap them and create new dll based on existing

This is well out of scope for what we'd want in our language and compiler.  I'd suggest if you want this, that you write your own tool that decompiles, updates, and then reemits all your references in this fashion.

**Reply by przemo098** [reply]
language is not for creators but for users, writing code should be as simple as possible even in case of rewriting fundamentals what happened with old .NET 

I'm trying to give idea not implementation

**Reply by HaloFour** [reply]
The idea is fine, it's just not something that the C# language compiler can or should do.  This is firmly in the territory of an assembly rewriter, which would be a tool external to the compiler.  That has a whole host of ramifications.

**Reply by CyrusNajmabadi** [reply]
> I'm trying to give idea not implementation

That's fine.  This is just not really the location for that discussion though.  It would need to go to some tool that is involved in dll rewriting.  Given that this is for testing purposes, i'd recommend a team responsible for testing/mocks/fakes/etc.

**Reply by iam3yal** [reply]
@przemo098 

> I think it wont be used often in app code, however it will simplify writing tests / overriding sealed classes in libs

It wouldn't simplify anything because it's already possible through various libraries like Moq, NSubstitute, AutoFixture that are completely free to use and there are many more tools.

The way to handle override methods in sealed classes is through wrappers and the tools I mentioned above makes it pretty easy if you understand how to use them and setup the tests correctly, it should certainly be a goal to make as less fakes as possible during testing and if you find yourself do it often then something might be off.

> What is really important in development lifecycle.

Testing is important but I doubt monkey patching as it exists in dynamic languages make sense in C# and even less so to facilitate testing.

**Reply by przemo098** [reply]
It's not for testing it's general approach what do you recommend than ?
Any team / repository in .NET area ?
also changing fundamentals will be usefull in case of destructuring

**Reply by przemo098** [reply]
@iam3yal 
Calculator calc = new CalculatorProxy(); // not var
you can have 
```
public void foo(Calculator calc){
    calc.Add();
}
```

**Reply by iam3yal** [reply]
@przemo098 Yes I noticed it later then I removed that part from my post. :)

**Reply by iam3yal** [reply]
@przemo098 

> It's **not for testing** it's general approach what do you recommend than ?

What do I recommend as opposed to monkey patching? or what do you mean?

You wrote the following in the OP:

> I think it **wont be used often in app code**, however **it will simplify writing tests / overriding sealed classes** in libs

So dunno. 🤷‍♂️

**Reply by przemo098** [reply]
@iam3yal it will be for sure used more than generators

**Comment by FaustVX** [comment]
Can this be implemented with Source Generator & Interceptors ?

**Reply by HaloFour** [reply]
To a point?  As soon as the target type is passed to another assembly outside of your control the interceptor would no longer be in play since the target type isn't actually changed in any way.  At best it gives the illusion that the immediate callsite in the test is overridden, but you could just as easily have done that by calling an extension method or something else entirely.

**Comment by colejohnson66** [comment]
The solution here is dependency injection. Use interfaces.

**Reply by DavidArno** [reply]
Partially. Writing sociable unit tests helps too as it removes the need for over-zealous isolation, which is likely behind the OP's request for a way to mock things that don't need mocking.

**Comment by huoyaoyuan** [comment]
This will invalidate one of the most important optimizations - inlining. Knowing something to be invariant is important, including method bodies, initialized-once values.

**Comment by TonyValenti** [comment]
Take a look at libHarmony. We use the at for C# monkey patching and it works great.

**Comment by InCerryGit** [comment]
You can not only try Harmony, but also give Jitex a try; it should be able to solve some of your problems.
https://github.com/Hitmasu/Jitex
